internal/controllers/machinedeployment/machinedeployment_status.go

```diff
@@ -0,0 +1,502 @@
+	if machineDeployment.Status.V1Beta2 == nil || machineDeployment.Status.V1Beta2.AvailableReplicas == nil {
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:   clusterv1.MachineDeploymentAvailableV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineDeploymentAvailableWaitingForAvailableReplicasSetV1Beta2Reason,
+		})
+		return
+	}
+
+	// minReplicasNeeded will be equal to md.Spec.Replicas when the strategy is not RollingUpdateMachineDeploymentStrategyType.
+	minReplicasNeeded := *(machineDeployment.Spec.Replicas) - mdutil.MaxUnavailable(*machineDeployment)
+
+	if *machineDeployment.Status.V1Beta2.AvailableReplicas >= minReplicasNeeded {
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:   clusterv1.MachineDeploymentAvailableV1Beta2Condition,
+			Status: metav1.ConditionTrue,
+			Reason: clusterv1.MachineDeploymentAvailableV1Beta2Reason,
+		})
+		return
+	}
+
+	message := fmt.Sprintf("%d available replicas, at least %d required", *machineDeployment.Status.V1Beta2.AvailableReplicas, minReplicasNeeded)
+	if machineDeployment.Spec.Strategy != nil && mdutil.IsRollingUpdate(machineDeployment) && machineDeployment.Spec.Strategy.RollingUpdate != nil {
+		message += fmt.Sprintf(" (spec.strategy.rollout.maxUnavailable is %s)", machineDeployment.Spec.Strategy.RollingUpdate.MaxUnavailable)
+	}
+	v1beta2conditions.Set(machineDeployment, metav1.Condition{
+		Type:    clusterv1.MachineDeploymentAvailableV1Beta2Condition,
+		Status:  metav1.ConditionFalse,
+		Reason:  clusterv1.MachineDeploymentNotAvailableV1Beta2Reason,
+		Message: message,
+	})
+}
+
+func setScalingUpCondition(_ context.Context, machineDeployment *clusterv1.MachineDeployment, machineSets []*clusterv1.MachineSet, bootstrapObjectNotFound, infrastructureObjectNotFound, getAndAdoptMachineSetsForDeploymentSucceeded bool) {
+	// If we got unexpected errors in listing the machine sets (this should never happen), surface them.
+	if !getAndAdoptMachineSetsForDeploymentSucceeded {
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:    clusterv1.MachineDeploymentScalingUpV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineDeploymentScalingUpInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	// Surface if .spec.replicas is not yet set (this should never happen).
+	if machineDeployment.Spec.Replicas == nil {
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:   clusterv1.MachineDeploymentScalingUpV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineDeploymentScalingUpWaitingForReplicasSetV1Beta2Reason,
+		})
+		return
+	}
+
+	desiredReplicas := *machineDeployment.Spec.Replicas
+	if !machineDeployment.DeletionTimestamp.IsZero() {
+		desiredReplicas = 0
+	}
+	currentReplicas := mdutil.GetReplicaCountForMachineSets(machineSets)
```

Comment:
This is using .spec.replicas. That doesn't seem correct to me

Reply:
Good catch!